x-pack/plugins/security_solution/server/lib/detection_engine/rules/bulk_edit_rules.ts

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { RulesClient } from '@kbn/alerting-plugin/server';
+import { BulkActionEditPayload } from '../../../../common/detection_engine/schemas/common';
+import { enrichFilterWithRuleTypeMapping } from './enrich_filter_with_rule_type_mappings';
+import { MlAuthz } from '../../machine_learning/authz';
+import { throwAuthzError } from '../../machine_learning/validation';
+
+import { ruleParamsModifier } from './bulk_edit/rule_params_modifier';
+import { splitBulkEditActions } from './bulk_edit/split_bulk_edit_actions';
+import { bulkEditActionToRulesClientOperation } from './bulk_edit/action_to_rules_client_operation';
+
+import { RuleAlertType } from './types';
+
+export interface BulkEditRulesArguments {
+  rulesClient: RulesClient;
+  actions: BulkActionEditPayload[];
+  filter?: string;
+  ids?: string[];
+  mlAuthz: MlAuthz;
+}
+
+/**
+ * calls rulesClient.bulkEdit
+ * transforms bulk actions payload into rulesClient compatible operations
+ * enriches query filter with rule types search queries
+ * @param BulkEditRulesArguments
+ * @returns edited rules and caught errors
+ */
+export const bulkEditRules = ({
```

Comment:
Could you point out where we ensure that prebuilt rules can't be edited?

Reply:
```typescript

        /**
         * validate rule params when rule is bulk edited (update and created in future as well)
         * returned params can be modified (useful in case of version increment)
         * @param mutatedRuleParams
         * @returns mutatedRuleParams
         */
        validateMutatedParams: (mutatedRuleParams) => {
          validateImmutable(mutatedRuleParams.immutable);
          validateIndexPatterns(mutatedRuleParams.index);

          return mutatedRuleParams;
        },
```

Reply:
It seems like this validation would be good to apply consistently across all RulesClient APIs that allow rules to be modified. However, when we update prebuilt rules we do want to be able to modify `immutable` rules. Is there a plan to be able to disable this validation in some cases so we can apply it to all user updates but not system updates to prebuilt rules?

Reply:
> It seems like this validation would be good to apply consistently across all RulesClient APIs that allow rules to be modified. However, when we update prebuilt rules we do want to be able to modify immutable rules. Is there a plan to be able to disable this validation in some cases so we can apply it to all user updates but not system updates to prebuilt rules?

For now it's applied only for bulkEdit

But validateMutatedParams has a following signature:
https://github.com/elastic/kibana/blob/a178595c450fe8d0ff34904b75d7a5795ec24ba6/x-pack/plugins/alerting/server/types.ts#L129

>   validateMutatedParams?: (mutatedOject: Params, origObject?: Params) => Params;

So, when using it for create, origObject will be empty. It can be taken in account and perform a different validation for that case

Reply:
> However, when we update prebuilt rules we do want to be able to modify immutable rules. Is there a plan to be able to disable this validation in some cases so we can apply it to all user updates but not system updates to prebuilt rules?

++ I think these are important questions.

I don't think that `validateMutatedParams` is a proper place for ensuring that prebuilt rules can't be modified. `origObject` won't be enough for doing this validation. At the rule type level we both allow and disallow creating/updating prebuilt rules:

- We allow creating or updating prebuilt rules from `addPrepackedRulesRoute`. It calls both `RulesClient.create` and `RulesClient.update`, and we pass `params.immutable: true` there.
- We disallow creating or updating prebuilt rules from all the other endpoints.

We'll need to find a proper place for doing this validation, but I'm ok with merging this PR as is.